src/main/java/com/iexec/worker/amnesia/AmnesiaRecoveryService.java

```diff
@@ -66,151 +66,105 @@ public AmnesiaRecoveryService(CustomFeignClient customFeignClient,
-
-                case REVEAL:
-                    shouldSubscribe = recoverReplicateByRevealing(contributionAuth);
-                    shouldKeepResultFolder = shouldSubscribe;
-                    break;
-
-                case UPLOAD_RESULT:
-                    shouldSubscribe = recoverReplicateByUploadingResult(contributionAuth);
-                    shouldKeepResultFolder = shouldSubscribe;
-                    break;
-
-                case COMPLETE:
-                    taskExecutorService.completeTask(chainTaskId);
-                    break;
+            if (!isResultAvailable && recoveryAction != RecoveryAction.CONTRIBUTE) {
+                log.error("Could not recover task, result not found [chainTaskId:{}, RecoveryAction:{}]",
+                        chainTaskId, recoveryAction);
+                continue;
             }
 
-            if (shouldSubscribe) {
-                subscriptionService.subscribeToTopic(chainTaskId);
-            }
+            Optional<AvailableReplicateModel> oReplicateModel =
+                    replicateService.retrieveAvailableReplicateModelFromContribAuth(contributionAuth);
 
-            if (shouldKeepResultFolder) {
-                recoveredChainTaskIds.add(chainTaskId);
+            if (!oReplicateModel.isPresent()) {
+                log.error("Could not recover task, no replicateModel retrieved [chainTaskId:{}, RecoveryAction:{}]",
+                        chainTaskId, recoveryAction);
+                continue;
             }
+
+            AvailableReplicateModel replicateModel = oReplicateModel.get();
+            recoverReplicate(interruptedReplicate, replicateModel);
+            recoveredChainTaskIds.add(chainTaskId);
         }
 
         return recoveredChainTaskIds;
     }
 
-    public boolean recoverReplicateByContributing(ContributionAuthorization contributionAuth) {
-        String chainTaskId = contributionAuth.getChainTaskId();
-
-        Optional<AvailableReplicateModel> oReplicateModel =
-                replicateService.retrieveAvailableReplicateModelFromContribAuth(contributionAuth);
+    public void recoverReplicate(InterruptedReplicateModel interruptedReplicate,
+                                 AvailableReplicateModel replicateModel) {
 
-        if (!oReplicateModel.isPresent()) {
-            log.info("could not retrieve replicateModel from contributionAuth to recover task "
-                    + "[chainTaskId:{}, RecoveryAction:CONTRIBUTE]", chainTaskId);
-            return false;
-        }
+        ContributionAuthorization contributionAuth = interruptedReplicate.getContributionAuthorization();
+        String chainTaskId = contributionAuth.getChainTaskId();
 
-        AvailableReplicateModel replicateModel = oReplicateModel.get();
+        switch (interruptedReplicate.getRecoveryAction()) {
+            case WAIT:
+                subscriptionService.subscribeToTopic(chainTaskId);
+                resultService.saveResultInfo(chainTaskId, replicateModel);
```

Comment:
Just wondering, couldn't we `resultService.saveResultInfo(chainTaskId, replicateModel);` once for all at the beginning of the `recoverReplicate(..)` method (so we don't miss any case)?

Reply:
Actually, we don't need it for all cases.